enhancements/kube-apiserver/startup-monitor.md

```diff
@@ -0,0 +1,234 @@
+
+### Goals
+
+- recover from a badly formatted or invalid pod manifest
+- recover from bad config observer output
+- recover from bad revisioned ConfigMaps and Secrets.
+
+### Non-Goals
+
+- recover from invalid non-revisioned certificates
+- recover from missing or invalid files on the masters given by static paths outside of the `/etc/kubernetes` directory (e.g. some certs or kubeconfig the operand is consuming)
+- optimize downtime of the API beyond in fallback case to O(60s). This mechanism is for the disaster case and only survival counts. If downtime is `O(5min)` or even more, this is completely ok.
+
+## Proposal
+
+We propose to add a `<operand>-startup-monitor` static pod that watches the operand pods for readiness (see below for details) and that is created by installer as another manifest in `/etc/kubernetes` and provides by the static pod controller the same way the operand pod manifest is created today.
+
+If the operand pod does not start up in N minutes, e.g. due to:
+
+- invalid pod manifest
+- free port wait loop timeout
+- crash-looping
+- not answering on the expected port (connection refused)
+- healthy but never readyz
+
+the task for startup-monitor is to fall back:
+
+1. when detecting problems with the new revision, the startup-monitor will copy the pod-manifest of the `/etc/kubernetes/static-pods/last-known-good` link (or the previous revision if the link does not exist, or don't do anything if there is no previous revision as in bootstrapping) into `/etc/kubernetes`.
+
+   It will add annotations to the old revision pod manifest:
+
+   - `startup-monitor.static-pods.openshift.io/fallback-for-revision: <revision>`
+   - `startup-monitor.static-pods.openshift.io/fallback-reason: MachineReadableReason`
+   - `startup-monitor.static-pods.openshift.io/fallback-message: "Human readable, possibly multiline message"`
+
+   These will be copied into the mirror pod by kubelet and the operator knows that this is due to a problem to start up the new revision. The message annotation will be published to the user through the operator degraded condition.
+
+2. if the operand becomes ready, the startup-monitor will link the revision to `/etc/kubernetes/static-pods/last-known-good`, and then remove its own pod-manifest from `/etc/kubernetes`, and hence commit suicide.
+
+As long as the startup-monitor notices an operand pod-manifest of a different revision (by checking the operator manifest in `/etc/kubernetes` and the revision annotation), it will do nothing and just wait, continuing to watch the operand pod-manifest. This is important to avoid races on startup, and avoid problems on downgrade before this mechanism was introduced.
+
+### Readiness
+
+The startup-monitor watches the mirror pod for readiness, i.e. by
+
+1. watch `/var/log/kube-apiserver/termination.log` for the first start-up attempt (beware of the race of startup-monitor startup and kube-apiserver startup). Set Reason=**NeverStartedUp** when this times out.
```

Comment:
what does it mean in practice? reading a line and comparing the content and timestamp?

Reply:
would it make sense to keep track of `pid` instead/additionally?